File: tools/palworld/src/transform.mjs
export function makeTransform(bounds, orientation, pixelW, pixelH) {
  const { min, max } = bounds;
  const { pxAxis, flipX, flipY } = orientation;
  const pyAxis = pxAxis === 'X' ? 'Y' : 'X';
  return (world) => {
    let x = ((world[pxAxis] - min[pxAxis]) / (max[pxAxis] - min[pxAxis])) * pixelW;
    let y = ((world[pyAxis] - min[pyAxis]) / (max[pyAxis] - min[pyAxis])) * pixelH;
    if (flipX) x = pixelW - x;
    if (flipY) y = pixelH - y;
    return { x, y };
  };
}

// Inverse of makeTransform: pixel (x, y) -> world { X, Y }. Used to emit raw
// world coordinates for cluster centroids (which are computed in pixel space).
export function makeInverseTransform(bounds, orientation, pixelW, pixelH) {
  const { min, max } = bounds;
  const { pxAxis, flipX, flipY } = orientation;
  const pyAxis = pxAxis === 'X' ? 'Y' : 'X';
  return (px, py) => {
    let fx = px;
    let fy = py;
    if (flipX) fx = pixelW - fx;
    if (flipY) fy = pixelH - fy;
    const world = { X: 0, Y: 0 };
    world[pxAxis] = (fx / pixelW) * (max[pxAxis] - min[pxAxis]) + min[pxAxis];
    world[pyAxis] = (fy / pixelH) * (max[pyAxis] - min[pyAxis]) + min[pyAxis];
    return world;
  };
}

File: tools/apps/palworld/maps/transform.py
"""World<->pixel coordinate transforms (port of transform.mjs)."""

from __future__ import annotations

from collections.abc import Callable

from .orientation import Orientation

# A world location is a mapping with "X"/"Y"/"Z" keys.
World = dict


def make_transform(bounds: dict, o: Orientation, pixel_w: float, pixel_h: float) -> Callable[[World], tuple[float, float]]:
    mn, mx = bounds["min"], bounds["max"]
    px_axis = o.px_axis
    py_axis = "Y" if px_axis == "X" else "X"

    def to_pixel(world: World) -> tuple[float, float]:
        x = (world[px_axis] - mn[px_axis]) / (mx[px_axis] - mn[px_axis]) * pixel_w
        y = (world[py_axis] - mn[py_axis]) / (mx[py_axis] - mn[py_axis]) * pixel_h
        if o.flip_x:
            x = pixel_w - x
        if o.flip_y:
            y = pixel_h - y
        return x, y

    return to_pixel


def make_inverse_transform(bounds: dict, o: Orientation, pixel_w: float, pixel_h: float) -> Callable[[float, float], World]:
    """Inverse of make_transform: pixel (x, y) -> world {X, Y}. Used to emit raw
    world coordinates for cluster centroids (computed in pixel space)."""
    mn, mx = bounds["min"], bounds["max"]
    px_axis = o.px_axis
    py_axis = "Y" if px_axis == "X" else "X"

    def to_world(px: float, py: float) -> World:
        fx, fy = px, py
        if o.flip_x:
            fx = pixel_w - fx
        if o.flip_y:
            fy = pixel_h - fy
        world = {"X": 0.0, "Y": 0.0}
        world[px_axis] = fx / pixel_w * (mx[px_axis] - mn[px_axis]) + mn[px_axis]
        world[py_axis] = fy / pixel_h * (mx[py_axis] - mn[py_axis]) + mn[py_axis]
        return world

    return to_world
